operators/src/engine/execution_context.rs

```diff
@@ -1,46 +1,163 @@
-use geoengine_datatypes::{
-    primitives::Coordinate2D,
-    raster::{GridShape2D, TilingSpecification},
-};
-
-use crate::concurrency::{ThreadPool, ThreadPoolContext};
+use crate::concurrency::ThreadPool;
+use crate::engine::{QueryRectangle, VectorResultDescriptor};
+use crate::error::Error;
+use crate::mock::MockDataSetDataSourceLoadingInfo;
+use crate::source::OgrSourceDataset;
+use crate::util::Result;
+use geoengine_datatypes::dataset::DataSetId;
+use geoengine_datatypes::raster::GridShape;
+use geoengine_datatypes::raster::TilingSpecification;
+use std::any::Any;
+use std::collections::HashMap;
+use std::fmt::Debug;
 use std::path::PathBuf;
 
-/// A context that provides certain utility access during operator execution
-/// TODO: maybe this is rather a trait
-#[derive(Debug, Clone)]
-pub struct ExecutionContext<'pool> {
-    pub raster_data_root: PathBuf,
-    pub thread_pool: ThreadPoolContext<'pool>,
-    pub tiling_specification: TilingSpecification,
+/// A context that provides certain utility access during operator initialization
+pub trait ExecutionContext:
+    Send
+    + Sync
+    + LoadingInfoProvider<MockDataSetDataSourceLoadingInfo, VectorResultDescriptor>
+    + LoadingInfoProvider<OgrSourceDataset, VectorResultDescriptor>
```

Comment:
> jdroenner commented on 20.01.2021
We talked about this. I don't think it is a good idea to add all loading provider requirements here. Just require them in the operator and implement them for the context?
>
> this also requires to change the box in initialize to T. Then in an operator we can do where T: LoadingInfoProvider<MeData, MeResult>

I see it differently. Any impl of ExecutionContext needs to be LoadingInfoProvider for all Operators because it has to give LoadingInfos to all operators. That's why I demand it here. If a operator needs only one specific LoadingInfoProvider it can ask for this one, but any implementation of ExecutionContext must provide them all

Reply:
There are pros and cons:

Pro:
When you implement an ExecutionContext, you know what to implement.

Cons:
Currently, you can add a new operator by just having it in the program and type-tagged. Now, you are restricted to a few loading providers.
You could do it in the `Any` fashion and try to resolve it.
This makes it extensible, but you can't type check it.

Reply:
I get the extensiblity argument. However, the `initialize` method of `Operator` only requires an `ExecutionContext`. How does the `GdalSource` e.g. know that the `&dyn ExecutionContext` also happens to be a `&dyn LoadingInfoProvider<Gdal..>`?

Reply:
You wouldn't, you just try to get it.

Reply:
Well i guess it does not matter if we add all the types here. To avoid a long discussion i guess we should add them. However, i guess we loose type safety anyway. The current approach already uses downcast, as i understand it, there is no way to have the compiler derive all the needed implementations? So you can implement a new source with a special provider without adding the LoadingInfoProvider<X,Y> to ExecutionContext.

Reply:
I only use downcast in the mock execution context to make a convenient wrapper.

Can we resolve this discussion (for now)?

Reply:
yes keep it as it is for now